examples/server/index.html.cpp

```diff
@@ -0,0 +1,278 @@
+    }
+
+    // send message to server
+    const chat = async (msg) => {
+      if (controller.value) {
+        console.log('already running...');
+        return;
+      }
+      controller.value = new AbortController();
+
+      const history = [...transcript.value, ['{{user}}', msg]];
+      transcript.value = history;
+
+      let additionalParams = {
+        message: msg,
+        history: history.flatMap(([name, msg]) => `${name}: ${msg}`).join("\n"),
+      }
+
+      const payload = template(chatTemplate.value, additionalParams)
+
+      let currentMessage = "";
+      await fetchEventSource('/completion', {
+        method: 'POST',
+        signal: controller.value.signal,
+        body: JSON.stringify({
+          stream: true,
+          prompt: payload,
+          n_predict: parseInt(nPredict.value),
+          temperature: parseFloat(temperature.value),
+          stop: ["</s>", template("{{bot}}:"), template("{{user}}:")]
+        }),
+        onmessage(e) {
+          const data = JSON.parse(e.data);
+          currentMessage += data.content;
+
+          if (data.stop) {
+            console.log("-->", data, ' response was:', currentMessage, 'transcript state:', transcript.value);
+          }
+
+          transcript.value = [...history, ['{{bot}}', currentMessage]]
+          return true;
+        },
+        onclose(e) {
+          controller.value = null;
+          return false;
+        },
+      });
+    }
+
+    function MessageInput() {
+      const message = useSignal("")
+
+      const stop = (e) => {
+        e.preventDefault();
+        if (controller.value) {
+          controller.value.abort();
+          controller.value = null;
+        }
+      }
+
+      const reset = (e) => {
+        stop(e);
+        transcript.value = [];
+      }
+
+      const submit = (e) => {
+        stop(e);
+        chat(message.value);
+        message.value = "";
+      }
+
+      return html`
+        <form onsubmit=${submit}>
+          <input type="text" value="${message}" oninput=${(e) => message.value = e.target.value} autofocus placeholder="Chat here..."/>
+          <button type="submit" disabled=${!generating.value} >Send</button>
+          <button onclick=${(e) => stop(e)} disabled=${generating.value}>Stop</button>
+          <button onclick=${(e) => reset(e)}>Reset</button>
+        </form>
+      `
+    }
+
+    const ChatLog = (props) => {
+      const messages = transcript.value;
+      const container = useRef(null)
+
+      useEffect(() => {
+        // scroll to bottom (if needed)
+        if (container.current && container.current.scrollHeight <= container.current.scrollTop + container.current.offsetHeight + 100) {
+          container.current.scrollTo(0, container.current.scrollHeight)
+        }
+      }, [messages])
+
+      const chatLine = ([user, msg]) => {
+        return html`<p><strong>${template(user, {})}:</strong> ${template(msg, {})}</p>`
```

Comment:
If messages are ever received out-of-order, you'll want to add a key (you can just use the msg object itself). If they're always appended, the key has no effect and could be left out.

```suggestion
        return html`<p key=${msg}><strong>${template(user, {})}:</strong> ${template(msg, {})}</p>`
```